dev-docs/RFCs/vNext/layer-extension-rfc.md

```diff
@@ -0,0 +1,157 @@
+# RFC: Layer Extension
+
+* **Authors**: Xiaoji Chen
+* **Date**: Dec 2018
+* **Status**: Draft
+
+
+## Overview
+
+This is a proposal for adding/removing optional functionalities to a deck.gl layer on-demand.
+
+
+## Motivation
+
+As the deck.gl layers ecosystem grows, we see many feature requests as variations of some existing layer. One of deck.gl's core design goals is easily composible and extensible layers. Over the last few major releases we introduced features such as [Composite Layer](/docs/api-reference/composite-layer.md), [shader modules](https://github.com/uber/luma.gl/blob/master/docs/developer-guide/shadertools/using-shader-modules.md), [shader injection](https://github.com/uber/luma.gl/blob/master/dev-docs/RFCs/v6.0/shader-fragment-injection-rfc.md), etc. While they offer much flexibility to authors of custom layers, it is not quite accessible to the majority of our users.
+
+To begin with, adding a feature to an existing core layer is no easy task. It usually involves modification to multiple steps in the layer lifecycle, e.g. adding new props, uniforms, declaring and generating new attributes, and injecting code into the shaders. It requires the author to understand the general WebGL render pipeline and GLSL, deck.gl's layer lifecycle and attribute management system, and sometimes luma.gl. This is outlined in our tutorial for [subclassed layers](/docs/developer-guide/subclassed-layers.md).
+
+Secondly, because of the ad-hoc nature of layer customization, it is difficult to package and reuse an "extension" for other layers. This is apprent in the [kepler.gl repo](https://github.com/uber/kepler.gl/tree/master/src/deckgl-layers) where a custom version of every core layer must be created in order to override a simple default bahavior.
+
+To mitigate this, we can of course add as many customizable functionalities to the base layer as possible. However, it is a major concern that inflating the base layer will sacrefice performance in return. Additional attributes takes CPU cycles to generate and memory to store. Larger shaders take longer to compile. WebGL1 has a restricting limit of 16 attributes per vertex shader, so adding attributes without the user asking for it will make the layer less extensible instead.
+
+
+## Case Studies
+
+### fp64
+
+* Shader module: `project64` (replaces `project32`)
```

Comment:
Not sure if we should keep this one (at least not in the RFC). Some observation:
- Does not explain how "replacement" happens (other cases are "additive")
- No details on what is being replaced (other cases have detailed lists).
- Still not convinced fp64 is worth touching at this point. 

A more interesting case would be general projections as injectable (replaceable?) shader modules.